docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,385 @@
+```
+
+Example with one service specifying IDs for know secrets and one without known secrets 
+
+```yaml
+ADD_SECRETSTORE: "appservice-xyz[Databases.application-service; <other TBD id>], appservice-http-export"
+```
+
+Once security-secretstore-setup generates the known application-service database credentials, it will save them in memory while processing additional adds of `SecretStores` so they can be used if other Application Services needs them seeded into their `SecretStores` . 
+
+### Abstraction Interface
+
+The following will be the new `SecretProvider` abstraction interface used by all Edgex services
+
+```go
+type SecretProvider interface {
+    // Stores new secrets into the service's exclusive SecretStore at the speficied path.
+	StoreSecrets(path string, secrets map[string]string) error
+	// Retrieves secrets from the service's exclusive SecretStore at the speficied path.
+	GetSecrets(path string, _ ...string) (map[string]string, error)
+    // Sets the secrets lastupdated time to current time. 
+    // Needed for InsecureSecrets support
+	SecretsUpdated()
+    // Returns the secrets last updated time
+	SecretsLastUpdated() time.Time
+}
+```
+
+> *Note: The `GetDatabaseCredentials` and `GetCertificateKeyPair` APIs have been removed. These are no longer needed since insecure database credentials will no longer be stored in the `DatabaseInfo` configuration and certificate key pairs are secrets like any others. This allows these secrets to be retrieved via the `GetSecrets` API.*
+
+### Implementation
+
+#### Factory Method and Bootstrap Handler
+
+The factory method and bootstrap handler will follow that currently in the Bootstrap implementation with some tweaks. Rather than putting the two split interfaces into the DIC, it will put just the single interface instance into the DIC. See details in the [Interfaces and factory methods](#interfaces-and-factory-methods) section above under **Existing Implementations**.
+
+#### Caching of Secrets
+
+Secrets will be cached as they are currently in the Application Service implementation
+
+#### Insecure Secrets
+
+Insecure Secrets will be handled as they are currently in the Application Service implementation. `DatabaseInfo` configuration will no longer be an option for storing the insecure database credentials. They will be stored in the `InsecureSecrets` configuration only.
+
+```toml
+[Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.DB]
+    	path = "redisdb"
+    	[Writable.InsecureSecrets.DB.Secrets]
+            username = ""
+            password = ""
+```
+
+##### Handling on-the-fly changes to `InsecureSecrets`
+
+All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap` as along with existing log level processing. This special processing will be taken from App SDK.
```

Comment:
```suggestion
All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap`  along with existing log level processing. This special processing will be taken from App SDK.
```